qiskit/transpiler/passes/routing/algorithms/bip_model.py

```diff
@@ -0,0 +1,393 @@
+
+        Raises:
+            TranspilerError: if unknow objective type is specified.
+        """
+        mdl = Model()
+
+        # *** Define main variables ***
+        # Add w variables
+        w = {}
+        for t in range(self.depth):
+            for q in range(self.num_vqubits):
+                for j in range(self.num_pqubits):
+                    w[t, q, j] = mdl.binary_var(name=f"w_{t}_{q}_{j}")
+        # Add y variables
+        y = {}
+        for t in range(self.depth):
+            for (p, q) in self.gates[t]:
+                for (i, j) in self._arcs:
+                    y[t, p, q, i, j] = mdl.binary_var(name=f"y_{t}_{p}_{q}_{i}_{j}")
+        # Add x variables
+        x = {}
+        for t in range(self.depth - 1):
+            for q in range(self.num_vqubits):
+                for i in range(self.num_pqubits):
+                    x[t, q, i, i] = mdl.binary_var(name=f"x_{t}_{q}_{i}_{i}")
+                    for j in self._coupling.neighbors(i):
+                        x[t, q, i, j] = mdl.binary_var(name=f"x_{t}_{q}_{i}_{j}")
+
+        # *** Define main constraints ***
+        # Assignment constraints for w variables
+        for t in range(self.depth):
+            for q in range(self.num_vqubits):
+                mdl.add_constraint(
+                    sum(w[t, q, j] for j in range(self.num_pqubits)) == 1,
+                    ctname=f"assignment_vqubits_{q}_at_{t}",
+                )
+        for t in range(self.depth):
+            for j in range(self.num_pqubits):
+                mdl.add_constraint(
+                    sum(w[t, q, j] for q in range(self.num_vqubits)) == 1,
+                    ctname=f"assignment_pqubits_{j}_at_{t}",
+                )
+        # Each gate must be implemented
+        for t in range(self.depth):
+            for (p, q) in self.gates[t]:
+                mdl.add_constraint(
+                    sum(y[t, p, q, i, j] for (i, j) in self._arcs) == 1,
+                    ctname=f"implement_gate_{p}_{q}_at_{t}",
+                )
+        # Gate can be implemented iff both of its qubits are located at the associated nodes
+        for t in range(self.depth):
+            for (p, q) in self.gates[t]:
+                for (i, j) in self._arcs:
+                    # Apply McCormick to y[t, p, q, i, j] == w[t, p, i] * w[t, q, j]
+                    mdl.add_constraint(
+                        y[t, p, q, i, j] >= w[t, p, i] + w[t, q, j] - 1,
+                        ctname=f"McCormickLB_{p}_{q}_{i}_{j}_at_{t}",
+                    )
+                    mdl.add_constraint(
+                        y[t, p, q, i, j] <= w[t, p, i],
+                        ctname=f"McCormickUB1_{p}_{q}_{i}_{j}_at_{t}",
+                    )
+                    mdl.add_constraint(
+                        y[t, p, q, i, j] <= w[t, q, j],
+                        ctname=f"McCormickUB2_{p}_{q}_{i}_{j}_at_{t}",
+                    )
```

Comment:
There is a better version of these constraints that is stronger.

Suggested:
```       
    for t in range(self.depth-1):
            for (p, q) in self.gates[t]:
                for (i, j) in self._arcs:
                    # Stronger version of McCormick: gate (p,q) is implemented at (i, j) iff i moves to i or j, and j moves to i or j
                    mdl.add_constraint(
                        y[t, p, q, i, j] >= x[t, p, i, i] + x[t, p, i, j] + x[t, q, j, i] + x[t, q, j, j] - 1,
                        ctname=f"McCormickLB_{p}_{q}_{i}_{j}_at_{t}",
                    )
                    mdl.add_constraint(
                        y[t, p, q, i, j] <= x[t, p, i, i] + x[t, p, i, j],
                        ctname=f"McCormickUB1_{p}_{q}_{i}_{j}_at_{t}",
                    )
                    mdl.add_constraint(
                        y[t, p, q, i, j] <= x[t, q, j, i] + x[t, q, j, j],
                        ctname=f"McCormickUB2_{p}_{q}_{i}_{j}_at_{t}",
                    )
    # For last time step, use regular McCormick
    for (p, q) in self.gates[self.depth - 1]:
           for (i, j) in self._arcs:
               # Apply McCormick to y[self.depth - 1, p, q, i, j] == w[self.depth - 1, p, i] * w[self.depth - 1, q, j]
               mdl.add_constraint(
                    y[self.depth - 1, p, q, i, j] >= w[self.depth - 1, p, i] + w[self.depth - 1, q, j] - 1,
                    ctname=f"McCormickLB_{p}_{q}_{i}_{j}_at_last",
                )
                mdl.add_constraint(
                    y[self.depth - 1, p, q, i, j] <= w[self.depth - 1, p, i],
                    ctname=f"McCormickUB1_{p}_{q}_{i}_{j}_at_last",
                )
                mdl.add_constraint(
                    y[self.depth - 1, p, q, i, j] <= w[self.depth - 1, q, j],
                    ctname=f"McCormickUB2_{p}_{q}_{i}_{j}_at_last",
                )
```

Reply:
I prefer the current constraints for several reasons:
- they are the same as described in arxiv paper cited in the docstring (users would read this code with the paper in hand and would be confused if different-looking constraints show up)
- they are fewer lines and easier to read (suggesting easier to understand)
- there is no difference in performance between the suggested ones and the current ones (at least QV circuits in my env).

Reply:
- I can update the paper, and I am planning to do so -- the reason they are not indicated in the paper is oversight, not a deliberate choice.
- True, the current constraints are easier to understand. 
- The proposed set of constraints is stronger. It is very well possible that one cannot observe the difference on smaller instances, as you say, but on larger QV circuits solved to optimality, the version that I am proposing is better. From an optimization standpoint, there is no reason to choose the weaker version. For large QV experiments, we need the best model we can get.

Reply:
I'd personally like to push the suggested change to a follow-up PR after the revised version of the paper is published in the future. I'd like to ask comments for others.

Reply:
If there's a better/stronger version, let's update to it (follow-up PR is fine).